configs: Update configs in configs/example/gem5_library for v25.0

erin-le opened the pull request.

erin-le committed:
configs: First pass to update gem5_library configs for v25.0

This commit contains the first pass to update the configs in
configs/example/gem5_library for gem5 v25.0. It updates the resources
to their latest versions and updates the comments accordingly. It
also removes the generator style of exit handling and adds
hypercall exit handlers where necessary.

erin-le committed:
configs: remove print statements at end of gem5_library configs

This commit removes print statements which are printed at the end of
gem5 library config scripts after simulation completes.

erin-le committed:
configs: update copyright dates in configs/example/gem5_library

erin-le committed:
configs: update spatter-gen-test.py

powerjg commented on configs/example/gem5_library/arm-demo-ubuntu-run.py:
```diff
@@ -45,7 +45,8 @@
 
-# This runs a check to ensure the gem5 binary interpreting this file is compiled to include the ARM ISA.
+# This runs a check to ensure the gem5 binary interpreting this file is
+# compiled to include the ARM ISA.
```
I don't actually think we need these checks in the example files. I believe that it will have duplicate checks when importing the `ArmBoard`. Let me know if I'm wrong.

erin-le replied:
It looks like the demo boards and `ArmBoard` have the `requires(...)` check, but `RiscvBoard` and `X86Board` use `if processor.get_isa() != ISA.[ISA]` instead. I'll remove the `requires(...)` check from the configs that use the demo boards and `ArmBoard`.

Should I also remove the `requries` check for configs that use `RiscvBoard` and `X86Board`?

powerjg replied:
It would be good for it to be consistent. I think it's best in the python/gem5/boards/*.py file, but that's just me.

powerjg commented on configs/example/gem5_library/arm-ubuntu-run.py:
```diff
@@ -105,27 +108,45 @@
+# default, or inherit directly from ExitHandler and specify a hypercall number.
+# See src/python/gem5/simulate/exit_handler.py for more information on which
+# behaviors map to which hypercalls, and what the default behaviors are.
+class CustomKernelBootedExitHandler(KernelBootedExitHandler):
```
A *HUGE* thank you for doing this!!

powerjg commented on configs/example/gem5_library/checkpoints/riscv-hello-restore-checkpoint.py:
```diff
@@ -87,6 +87,14 @@
 # was taken from running this gem5 configuration script,
 # configs/example/gem5_library/checkpoints/riscv-hello-save-checkpoint.py
+
+# gem5 v25.0: `riscv-hello` on gem5 resources needs to be updated to say it
+# is compatible with v25.0
+# The description of the riscv-hello-example-checkpoint says it is only
+# compatible with gem5 v22.0, but the raw JSON shows that it is compatible
+# with gem5 v24.1 and v25.0
```
Is this a "TODO"? If so, make an issue.

erin-le replied:
Issue with all TODOs from this PR here: https://github.com/gem5/gem5/issues/2453

powerjg commented on configs/example/gem5_library/checkpoints/riscv-hello-save-checkpoint.py:
```diff
@@ -92,30 +96,36 @@
 max_ticks = 10**6
-simulator = Simulator(board=board, full_system=False, max_ticks=max_ticks)
-simulator.run()
+m5.scheduleTickExitAbsolute(max_ticks)
```
Two things:
1. Why are we using `m5.scheduleTickAbsolute`? Doesn't the `Simulator` have a function for this?
2. Use `1_000_000` instead of a floating point number.

erin-le replied:
Due to the [comment in the hypercall 6 handler](https://github.com/gem5/gem5/blob/654fa090c17c82eeaefff7435929528e858afcf6/src/python/gem5/simulate/exit_handler.py#L184), I was always under the impression that only `m5.scheduleTickExitFromCurrent` or `m5.scheduleTickExitAbsolute` could be used to trigger a hypercall 6 exit. Can `Simulator.schedule_max_insts()` also trigger hypercall 6? From the comment in `schedule_max_insts()`, it looks like it triggers an ExitEvent.MAX_INSTS exit event.

powerjg replied:
There's exits for both *instructions* and for *ticks*. Which one is the goal here?

There should be just one hypercall/exit number for all tick-based exits (or we should be consistent). So, whatever we use in the simulator should be consistent.

We never want to use the `m5` module in these example standard library scripts, if at all possible. If we need to add a new API to the simulator, we should do that.

erin-le replied:
> There's exits for both instructions and for ticks. Which one is the goal here?

Hey Jason, sorry for the mixup; I should have been looking at `Simulator.set_max_ticks()` instead of `Simulator.schedule_max_insts()`. 

> We never want to use the m5 module in these example standard library scripts, if at all possible. If we need to add a new API to the simulator, we should do that.

We will probably need to add another API to Simulator. Currently, it doesn't look like there's a function that allows access to `m5.scheduleTickExitFromCurrent` or `m5.scheduleTickExitAbsolute`.

powerjg replied:
Yes, we should add that API if it doesn't exist.

erin-le replied:
PR adding API here: https://github.com/gem5/gem5/pull/2459

powerjg commented on configs/example/gem5_library/checkpoints/simpoints-se-checkpoint.py:
```diff
@@ -110,33 +110,43 @@
+#     ),
+# )
+
+# The workload x86-print-this-15000-with-simpoints, version 2.0.0 is available
+# for gem5 versions 23.1 through 24.1. We may need to create another version
+# of the workload that throws a hypercall (see comment below)
```
Make an issue, please. If you want to go ahead and merge this PR, reference the issue here.

powerjg commented on configs/example/gem5_library/checkpoints/simpoints-se-restore.py:
```diff
@@ -106,58 +102,76 @@
-simulator.schedule_max_insts(board.get_simpoint().get_warmup_list()[0])
+# Here, m5.scheduleTickExitAbsolute schedules an exit event for the first
+# SimPoint's warmup instructions
+m5.scheduleTickExitAbsolute(board.get_simpoint().get_warmup_list()[0])
```
Again, why not use `schedule_max_insts()`?

erin-le replied:
I misread `insts` as `ticks` in the original function and tried switching it to m5.scheduleTickExitAbsolute as a result. We will probably have to modify the Simulator API if we want to use hypercalls for this, as `schedule_max_insts` results in a MAX_INSTS classic exit event.

powerjg replied:
I think it would be OK to update the simulator to use the "new" event as long as we make a note of it in the release notes

erin-le replied:
I opened a draft PR trying to update `schedule_max_insts()` to use hypercalls here: https://github.com/gem5/gem5/pull/2514, but it doesn't work as intended.

powerjg commented on configs/example/gem5_library/exit_handling/user-exit-handler.py:
```diff
@@ -72,17 +71,19 @@
 )
 
 # Set the workload
+# The binary on gem5 resources needs to be updated to say that it is
+# compatible with gem5 v25.0
```
To do?

powerjg commented on configs/example/gem5_library/looppoints/create-looppoint-checkpoints.py:
```diff
@@ -117,6 +117,10 @@
 dir = Path(args.checkpoint_path)
 dir.mkdir(exist_ok=True)
 
+# The workload may have to be updated so it throws a hypercall instead of
+# hypercall 0. Then, this config will have to be modified so it has a handler
+# for that hypercall
```
Issue?

powerjg commented on configs/example/gem5_library/looppoints/restore-looppoint-checkpoint.py:
```diff
@@ -115,6 +116,7 @@
     cache_hierarchy=cache_hierarchy,
 )
 
+# This workload may have to be updated to throw hypercalls
```
To do? Issue?

powerjg commented on configs/example/gem5_library/memory_traffic.py:
```diff
@@ -99,6 +99,9 @@ def generator_factory(
 # tasks. Can replace the cache_hierarchy with any hierarchy to simulate the
 # cache as well as the memory
+
+# Need to test this; will this work with cache_hierarchy as None instead of
+# NoCache?
```
The test board should know how to deal with `None` cache instead of `NoCache`.  `NoCache` is actually a cache hierarchy, but that doesn't have any caches. `None` actually means that the generators are directly connected to the memory.

powerjg commented on configs/example/gem5_library/riscv-fs.py:
```diff
@@ -51,6 +51,10 @@
 from gem5.simulate.simulator import Simulator
 from gem5.utils.requires import requires
 
+# Will we continue to support the `riscv-disk-img` resource? If not, should this
+# config script be deprecated, given that it would be very similar to
+# `riscv-ubuntu-run.py` if we updated the workload to a more recent one?
```
Yes. I think this should be deprecated. That should be a separate PR, though.

erin-le replied:
Link to PR to deprecate `riscv-fs.py`: https://github.com/gem5/gem5/pull/2452

powerjg commented on configs/example/gem5_library/riscv-rvv-example.py:
```diff
@@ -38,14 +39,15 @@
 
 """
-
+# Should I add the UC Davis copyright if I modify this? Or do changes have to
+# be above a certain size/amount to add the copyright?
```
Use your best judgement. In this case, I don't think it matters.

powerjg commented on configs/example/gem5_library/riscvmatched-fs.py:
```diff
@@ -59,28 +65,36 @@
     l2_size="2MiB",
     is_fs=True,
 )
+# We previously tried to update this config script to boot Ubuntu 24.04, but
+# reverted back to Ubuntu 20.04 because we ran into issues. This will have
+# to be tested again.
```
To do or create an issue.

powerjg commented on configs/example/gem5_library/x86-gapbs-benchmarks.py:
```diff
@@ -64,19 +61,28 @@
+    description="An example configuration script to run the GAPBS benchmarks."
 )
 
+# This suite on the gem5 Resources website should be updated to use the latest
+# GAPBS benchmarks. For now, I will modify this script not to use the suite
```
To do or issue?

erin-le commented on configs/example/gem5_library/memory_traffic.py:
```diff
@@ -99,6 +99,9 @@ def generator_factory(
+# NoCache?
 board = TestBoard(
     clk_freq="1GHz",  # Ignored for these generators
     generator=generator,  # We pass the traffic generator as the processor.
```
Need to test this; will it work with `cache_hierarchy=None` instead of `cache_hierarchy=NoCache()`?

powerjg replied:
It should work with both.

powerjg commented on configs/example/gem5_library/x86-gapbs-benchmarks.py:
```diff
@@ -139,40 +145,72 @@
 # Here we set the FS workload, i.e., gapbs benchmark program
-# After simulation has ended you may inspect
-# `m5out/system.pc.com_1.device` to the stdout, if any.
+# After simulation has ended you may inspect `m5out/system.pc.com_1.device`
```
I think this path is wrong. I think it should be `m5out/board.pc.com_1.device`

powerjg commented on configs/example/gem5_library/x86-gapbs-benchmarks.py:
```diff
@@ -139,40 +145,72 @@
+        return True
+
+
+# def handle_workbegin():
+#     print("Done booting Linux")
+#     print("Resetting stats at the start of ROI!")
+#     m5.stats.reset()
+#     global start_tick
+#     start_tick = m5.curTick()
+#     processor.switch()
+#     yield False  # E.g., continue the simulation.
+
+
+# def handle_workend():
+#     print("Dump stats at the end of the ROI!")
+#     m5.stats.dump()
+#     yield True  # Stop the simulation. We're done.
```
delete?

powerjg commented on configs/example/gem5_library/x86-gapbs-benchmarks.py:
```diff
@@ -139,40 +145,72 @@
-    },
-)
+        m5.stats.dump()
+        # This is a hacky way to keep the script's ending print statement.
```
Don't worry about this. It can be dropped.

powerjg commented on configs/example/gem5_library/x86-global-inst-tracker.py:
```diff
@@ -113,6 +113,7 @@
     memory=memory,
     cache_hierarchy=cache_hierarchy,
 )
+# The workload needs to be updated to use hypercalls
```
Issue?

powerjg commented on configs/example/gem5_library/x86-npb-benchmarks.py:
```diff
@@ -99,14 +106,15 @@
-    help="Optionally put the maximum number of ticks to execute during the "
-    "ROI. It accepts an integer value.",
-)
+# parser.add_argument(
+#     "--ticks",
+#     type=int,
+#     help="Optionally put the maximum number of ticks to execute during the "
+#     "ROI. It accepts an integer value."
+# )
```
I think you can delete all of the commented code in this file.

powerjg commented on configs/example/gem5_library/x86-parsec-benchmarks.py:
```diff
@@ -63,6 +63,11 @@
 from gem5.simulate.simulator import Simulator
 from gem5.utils.requires import requires
 
+# It doesn't look like we have any up to date Parsec resources. The
```
Deprecate this script.

powerjg reviewed (commented):
First.. **amazing job!!** You have done a great job of cleaning this up. Way to go 🥇 

Have all (any) of these been tested? We should figure out a way to communicate what has been tested.

Also, everything that is something to do in the future should either be marked as "to do" before this is merged or an issue should be created (and there should be cross links from the issue to the code and vice versa).

erin-le commented:
@powerjg: Hey Jason, thanks for the review. I'm still in the progress of cleaning up these scripts, so there will be more changes shortly.

None of these scripts have been tested yet. I'll also extract the comments about things that need to be updated into an issue, as you mentioned above.

erin-le committed:
configs: add more small changes; further changes still to be made

erin-le committed:
configs: deprecate x86-parsec-benchmarks.py

This commit deprecates x86-parsec-benchmarks.py in
configs/example/gem5_library, as we don't have any up to date PARSEC
resources.

erin-le committed:
configs: remove `requires(...)` check from certain configs

This commit removes the `requires(...)` check from config scripts
that use the demo and Arm boards, as these boards already perform
the check.

erin-le committed:
configs: Add custom hypercall 1-3 handlers to ubuntu run scripts

This commit adds custom hypercall 1-3 hypercall handlers to example
configs that boot Ubuntu and use the ArmBoard, RiscvBoard, or X86Board.

erin-le committed:
configs: remove work items from comments

This commit removes comments that were made as temporary notes to
record future work items needed to bring the config scripts up
to date for gem5 v25.0.

erin-le committed:
configs: remove requires(isa_required) from x86 and riscv configs

This commit removes the `requires(isa_required= ... )` check from
X86 and RISCV configs to be consistent with configs that use the
demo boards and ArmBoard.

erin-le committed:
configs: fix path to simulated terminal output

Previously, some configs in configs/example/gem5_library had
incorrect paths to the file in m5out containing a simulated system's
terminal output. This commit corrects these paths.

erin-le committed:
configs: modify x86-gapbs-benchmarks.py

This commit removes/modifies comments and removes unused code from
x86-gapbs-benchmarks.py.

erin-le committed:
configs: modify x86-npb-benchmarks.py

This commit modifies x86-npb-benchmarks.py to remove comments and
unused code. It also modifies it to obtain NPB workloads directly,
instead of using a suite, as there currently isn't a NPB suite
that uses up to date NPB workloads.

erin-le committed:
configs: deprecate configs/example/gem5_library/riscv-fs.py

This commit deprecates configs/example/gem5_library/riscv-fs.py,
as we no longer support riscv-disk-img for gem5 v25.0, and
updating the disk image would make riscv-fs.py nearly identical
to riscv-ubuntu-run.py.

erin-le committed:
configs: remove unused imports from gem5_library configs

erin-le committed:
Merge branch 'develop' into update-configs-example-v25.0

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

erin-le committed:
configs: update config to use new Simulator hypercall 6 API

This commit updates riscv-hello-save-checkpoint.py to use the new
Simulator hypercall 6 API instead of importing m5 and using
m5.scheduleTickExitAbsolute().

erin-le committed:
configs: restore simpoints-se-restore.py to original state

This config script was mistakenly updated to exit after reaching
a certain number of ticks instead of instructions. An
API to trigger a hypercall exit after reaching a certain number of
instructions currently does not exist, so this script is being
reset to how it previously was.

erin-le committed:
configs: second pass to update configs/example/gem5_library

This commit consists of changes made to the scripts in
configs/example/gem5_library after testing the previous sets of
changes. It also includes changes made after further proof-reading
of comments.

erin-le committed:
configs: additional changes to comments

This commit edits the comments in simpoints-se-restore.py, adds
a comment in the dramsys configs saying that they need to be
run from the base `gem5` directory, and adds comments to
`x86-global-inst-tracker.py` and
`looppoints/create-looppoint-checkpoints.py` saying that they will
only run on an X86 host machine.

erin-le commented:
At this point, all of the configs except for `x86-mi300x-gpu.py`, `x86-spec-cpu2006-benchmarks.py`, and `x86-spec-cpu2017-benchmarks.py` have been tested. This PR is ready for review. Additional concerns I came across while testing have been recorded in a second comment in issue https://github.com/gem5/gem5/issues/2453.

powerjg commented:
It would be great if @Harshil2107 could review
